### PR TITLE
[Accessibility] Improve checkout error messages

### LINKED ISSUE #37602: Accessibility Improvements ‣ Cart | Checkout ‣ Errors and Feedback
As originally noted [over here](https://github.com/woocommerce/woocommerce/issues/24185#issuecomment-779227823), within the cart and checkout pages any errors that display are not programmatically associated with, or within proximity of, their associated fields.

<div align="center">error-association</div>

The potential impact is that users of screen readers will have a difficult time understanding why a form is being rejected, which fields are in error, and how to remedy them. Proposals for us to investigate:

- Ensure `aria-invalid="true"` is on invalid form fields and updated to a`aria-invalid="false"` or removed for valid fields.
- Ensure error messages are nested inside of the element.
- Ensure errors that can't be nested with the element are programmatically associated with the input. Add `aria-describedby` to the input element and point to the ID of the error message.
- Ensure focus moves to the first field that has an error.

This is a relatively large task and in respect of changing where errors display, may require some design input.

### PR DESCRIPTION
**Pull request by Manussakis**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This is the second PR dedicated to close the #37602 issue.
**First PR**: https://github.com/woocommerce/woocommerce/pull/48738

Closes #37602  .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR improves the form error messages on the classic checkout page.

1. Add a product to the cart.
2. Go to the classic checkout page.
3. Turn on the screen reader.
4. Submit the form with the required fields empty.
6. Verify a notice element appears at the top of the form. The notice element style may vary per theme.
image
7. Verify the notice element contains an error message for each one of the required fields with error.
8. Verify the focus went to the notice element and the screen reader announced its content.
9. Verify you can tab through the error messages inside the notice element.
10. The hover style of the links inside the notice element depends on the theme installed.
11. Verify each one of the error messages moves the focus to the associated form field once clicked.
14. Verify the fields with errors have an inline message below them.
15. By default, the style of the inline message is: `color: #a00;`,  `font-size: .75em;`, and `margin-bottom: 0;`. However, it can vary depending on the installed theme.
image
16. The inline messages should disappear once the field value is updated and not appear again until the form has been submitted. This behavior is because the error messages come translated from the server and can be different depending on the value of the input.
17. Perform the steps above logged in/out.
18. Perform the test above with the "Shipping to a different address" form opened.
19. The logic that validates the inputs on the client side when their value change remains the same.
20. Perform the tests against the WP default themes:
    - Twenty Seventeen
    - Twenty Nineteen
    - Twenty Twenty
    - Twenty Twenty-One
    - Twenty Twenty-Two
    - Twenty Twenty-Three
    - Twenty Twenty-Four

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/client/legacy/js/frontend/address-i18n.js
Comment (Manussakis) [inline plugins/woocommerce/client/legacy/js/frontend/address-i18n.js:20]: Prevent screen readers from announcing the asterisk and make them announce the term "required" instead.

File: plugins/woocommerce/client/legacy/js/frontend/address-i18n.js
Comment (Manussakis) [inline plugins/woocommerce/client/legacy/js/frontend/address-i18n.js]: Prevent screen readers from announcing the asterisk and make them announce the term "required" instead.

File: plugins/woocommerce/client/legacy/js/frontend/checkout.js
Comment (Manussakis) [inline plugins/woocommerce/client/legacy/js/frontend/checkout.js:38]: Run the `validate_field` callback when the focus moves away from the input.

File: plugins/woocommerce/client/legacy/js/frontend/checkout.js
Comment (Manussakis) [inline plugins/woocommerce/client/legacy/js/frontend/checkout.js:220]: Small refactor – replace the `else if` statement with `||`.

File: plugins/woocommerce/client/legacy/js/frontend/checkout.js
Comment (Manussakis) [inline plugins/woocommerce/client/legacy/js/frontend/checkout.js]: 1. Overall, `<ul>` elements must have the default `list` role (`role="list"`). This became a problem in this task because the list items have been wrapped by the `<a>` tag and Voice Over keeps announcing the content infinitely.
The solution was to remove the `role="alert"` attribute from the `<ul>` element and add it to the parent.

2. The `tabindex="-1"` attribute is needed so the live region can be focused once populated. This allows users to tab through the links inside it and go straight to the input they want to fix.

File: plugins/woocommerce/client/legacy/js/frontend/checkout.js
Comment (Manussakis) [inline plugins/woocommerce/client/legacy/js/frontend/checkout.js:579]: The `submit_error` method expects the message content as a string. The conversion is done by calling `prop( 'outerHTML' )`.

File: plugins/woocommerce/includes/class-wc-checkout.php
Comment (Manussakis) [inline plugins/woocommerce/includes/class-wc-checkout.php]: Add the `id` attribute to the "terms" error message so it can be associated with the input.

File: plugins/woocommerce/includes/wc-template-functions.php
Comment (Manussakis) [inline plugins/woocommerce/includes/wc-template-functions.php]: Prevent screen readers from announcing the asterisk and make them announce the term "required" instead.

File: plugins/woocommerce/tests/e2e-pw/tests/shopper/cart-checkout-calculate-tax.spec.js
Comment (Manussakis) [inline plugins/woocommerce/tests/e2e-pw/tests/shopper/cart-checkout-calculate-tax.spec.js]: The `getByLabel` method includes the asterisk even if it has the `aria-hidden="true"` attribute. That's why it hasn't been removed in this case.

**Comment by Manussakis** [comment]
@dkotter This is the second PR dedicated to closing the https://github.com/woocommerce/woocommerce/issues/37602 issue. It's ready for code review.

-----

For the record, this is the first PR: https://github.com/woocommerce/woocommerce/pull/48738

**Review by dkotter — APPROVED** [review]
@ankitguptaindia @qasumitbagthariya This is ready for QA

**Comment by qasumitbagthariya** [comment]
### QA Update ✅ 

-------------

 I have checked this issue in the `Manussakis:fix/37602-checkout-form-error-feedback` branch working as expected.


**Testing Environment** 

<Details>

- WordPress: 6.5.5
- Theme: Twenty Twenty-Four 1.1
- WooCommerce - 9.0.2
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 

</Details>

Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for code review (Woo)


https://github.com/woocommerce/woocommerce/assets/67687255/120e7764-584c-49a3-b028-afcef68f8639

**Comment by vikrampm1** [comment]
@Manussakis please remove [WIP] and make this PR ready for review.

@coreymckrill this PR is ready for Woo review.

**Comment by vikrampm1** [comment]
@senadir This PR is awaiting your review.

**Review by senadir — APPROVED** [review]
Thank you for working on this, I followed the testing steps logged in and not, as well as reviewing the code.

**Comment by jeffpaul** [comment]
@senadir is there anything additional needed from us before this is good to merge in / prepare for a Woo release?

**Comment by vikrampm1** [comment]
@senadir this PR has completed all the steps from our end and is ready to be merged by you / Woo.

**Comment by vikrampm1** [comment]
@senadir / @woocommerce/rubik please merge this PR as the associated PR was already merged and the linked issue is closed.

**Comment by opr** [comment]
@vikrampm1 this PR has merge conflicts. I'd recommend rebasing on WooCommerce trunk and we will re-run CI.

**Comment by Manussakis** [comment]
Several CI tests failed after merging trunk into this branch.
I will work on fixing it and leave a comment when it's ready for review again.

cc: @vikrampm1 .

**Comment by Manussakis** [comment]
I have removed the code that hides for screen readers the asterisk of the required input labels. 
[This PR](https://github.com/woocommerce/woocommerce/pull/49696) is already handling this issue. Therefore, I removed the following step from the QA instruction: 
> 12. Verify the asterisk is not announced by the screen reader when the field is focused ("Star").

@opr I have rebased on WooCommerce trunk.

**Comment by Manussakis** [comment]
@opr Please hold off on merging this PR. I realized it has an issue similar to [the one mentioned in this PR](https://github.com/woocommerce/woocommerce/pull/48738#pullrequestreview-2204533147).

I will add styles for the input inline error message to the default WP themes.

**Comment by Manussakis** [comment]
@dkotter This PR has undergone some changes.
Please check these two comments for more context:
- https://github.com/woocommerce/woocommerce/pull/49094#issuecomment-2313338965
- https://github.com/woocommerce/woocommerce/pull/49094#issuecomment-2313641067

Please re-review it. Thanks!

@qasumitbagthariya A new round of QA will be necessary once the changes are approved. Thanks!

**Review by dkotter — APPROVED** [review]

**Comment by Manussakis** [comment]
@qasumitbagthariya Please, test this PR again. Thanks!

**Comment by qasumitbagthariya** [comment]
### QA Update ✅ 

-------------
@Manussakis 

 I have checked this issue in the `Manussakis:fix/37602-checkout-form-error-feedback` branch working as expected.

Cross check with different theme like

1. Storefront
2. TT4

**Testing Environment** 

<Details>

- WordPress: 6.5.5
- Theme: Twenty Twenty-Four 1.1
- WooCommerce - 9.4.0-dev
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 

</Details>

https://github.com/user-attachments/assets/de08a3a4-47d4-45fd-b02c-281395d7ff20


https://github.com/user-attachments/assets/2bbb3196-4d2a-451c-be40-d0ecf7b55947

Update:
@vikrampm1 PR is ready for the next step.

**Comment by vikrampm1** [comment]
@opr this PR is ready for your re-review and then merging, once it is good to go.

**Comment by jeffpaul** [comment]
@senadir @opr are there any additional code review concerns here or is this ready to merge?

**Comment by opr** [comment]
Hi - I re-ran CI and all the jobs are passing besides ones that can't because it's a fork.

@senadir do you want to re-review? I see @dkotter has approved this PR already.

**Comment by senadir** [comment]
Looking into it, I missed several pings here.

**Review by senadir — APPROVED** [review]
Thank you! I tested this with a variety of themes and it's looking good.

I do have one question, when errors are not matching, only one of them is visible.

For example, when the form is empty, those are the visible errors:
image


If I insert a wrong email (`test.test`), those errors are no longer shown in the notice:
image


Previously visible errors messages are gone, but the field still has the red border.

Would that be outside scope for this PR?

**Comment by Manussakis** [comment]
@senadir Thanks for testing it.

This PR takes the checkout error messages and makes their UI accessible, but the logic to determine what the error messages should look like is out of its scope.

I followed the steps you listed to reproduce the issue on the trunk branch. As you can see in the screenshot, even having other fields with errors, only the email error message is returned, just like you flagged.

Screenshot 2024-10-03 at 19 03 44

Thanks!

**Comment by vikrampm1** [comment]
@senadir please review the above comment and proceed with merging the PR if everything looks good.

**Comment by vikrampm1** [comment]
@senadir just a reminder on the above.

**Comment by senadir** [comment]
Thank you! good with me for now.

**Pull request merged**
